optuna/samplers/_cmaes.py

```diff
@@ -422,40 +427,77 @@ def sample_relative(
         study._storage.set_trial_system_attr(
             trial._trial_id, generation_attr_key, optimizer.generation
         )
-        study._storage.set_trial_system_attr(trial._trial_id, "cma:n_restarts", n_restarts)
+        study._storage.set_trial_system_attr(
+            trial._trial_id, self._attr_keys.n_restarts, n_restarts
+        )
 
         external_values = trans.untransform(params)
 
         return external_values
 
+    @property
+    def _attr_keys(self) -> _CmaEsAttrKeys:
+
+        if self._use_separable_cma:
+            attr_prefix = "sepcma:"
+        elif self._with_margin:
+            attr_prefix = "cmawm:"
+        else:
+            attr_prefix = "cma:"
+
+        def generation_attr_key_template(restart: int) -> str:
+            if self._restart_strategy is None:
+                return attr_prefix + "generation"
+            else:
+                return attr_prefix + "restart_{}:generation".format(restart)
+
+        return _CmaEsAttrKeys(
+            attr_prefix + "optimizer",
```

Comment:
Since `_restore_optimizer` returns 0 for n restarts. I doubt this attr_key is doing anything, or even worse this function seems not to restore the optimizer correctly.
https://github.com/optuna/optuna/blob/8bef4de77284965a805560030f169eaa7efdcaf9/optuna/samplers/_cmaes.py#L502

Reply:
Is it OK for `attr_keys` not to contains `restart_n` information?

Reply:
> Since `_restore_optimizer` returns 0 for n restarts. I doubt this attr_key is doing anything, or even worse this function seems not to restore the optimizer correctly.

`return None, 0` means that the optimizer must be initialized. So the logic seems to be correct.

Reply:
I added some test cases to check this logic and it works expected even after restarted. Let me resolve this comment.
https://github.com/gen740/optuna/compare/cmaes-sys-attr-key...c-bata:optuna:cmaes-sys-attr-key?expand=1